cpp/src/io/json/data_casting.cuh

```diff
@@ -0,0 +1,357 @@
+        auto end_index   = in.second - (options.keepquotes ? 0 : 1);
+        for (decltype(in.second) i = start_index; i < end_index; ++i) {
+          // Previous char was an escape char
+          if (escape) {
+            // A unicode code point escape sequence is \uXXXX
+            auto constexpr NUM_UNICODE_ESC_SEQ_CHARS = 6;
+            // The escape sequence comprises four hex digits
+            auto constexpr NUM_UNICODE_ESC_HEX_DIGITS = 4;
+            // A name for the char following the current one
+            auto constexpr NEXT_CHAR = 1;
+            // A name for the char after the next char
+            auto constexpr NEXT_NEXT_CHAR = 2;
+            // Reset escape flag for next loop iteration
+            escape = false;
+
+            // Check the character that is supposed to be escaped
+            auto escaped_char = get_escape_char(in.first[i]);
+
+            // This is an escape sequence of a unicode code point: \uXXXX,
+            // where each X in XXXX represents a hex digit
+            if (escaped_char == UNICODE_SEQ) {
+              // Make sure that there's at least 4 characters left from the
+              // input, which are expected to be hex digits
+              if (i + NUM_UNICODE_ESC_HEX_DIGITS < end_index) {
+                auto hex_val = string_to_hex(&in.first[i + NEXT_CHAR]);
+                if (hex_val < 0) {
+                  sizes[row] = 0;
+                  clear_bit(null_mask, row);
+                  return;
+                }
+                // Skip over the four hex digits
+                i += NUM_UNICODE_ESC_HEX_DIGITS;
+
+                // If this may be a UTF-16 encoded surrogate pair:
+                // we expect another \uXXXX sequence
+                int64_t hex_low_val = 0;
+                if (i + NUM_UNICODE_ESC_SEQ_CHARS < end_index && in.first[i + NEXT_CHAR] == '\\' &&
+                    in.first[i + NEXT_NEXT_CHAR] == 'u') {
+                  hex_low_val = string_to_hex(&in.first[i + 3]);
+                }
+                // This is indeed a surrogate pair
+                if (hex_val >= 0xD800 && hex_low_val >= 0xDC00) {
+                  // Skip over the second \uXXXX sequence
+                  i += NUM_UNICODE_ESC_SEQ_CHARS;
+                  uint32_t unicode_code_point =
+                    0x10000 + (hex_val - 0xD800) + (hex_low_val - 0xDC00);
+                  auto utf8_chars = strings::detail::codepoint_to_utf8(unicode_code_point);
+                  out_size += strings::detail::bytes_in_char_utf8(utf8_chars);
+                }
+                // Just a single \uXXXX sequence
+                else {
+                  auto utf8_chars = strings::detail::codepoint_to_utf8(hex_val);
+                  out_size += strings::detail::bytes_in_char_utf8(utf8_chars);
+                }
+              } else {
+                sizes[row] = 0;
+                clear_bit(null_mask, row);
+                return;
```

Comment:
This block appears five times :D I'd love to remove the repetition, but I'm not sure what's the best option (without perf overhead). Thoughts?

Reply:
Yeah, wasn't a fan of that either. I refactored it big times now. I think it's much cleaner now. No more code duplication for `computing output string size` versus `writing output string`.